### PR TITLE
Eliminate leak in merge_ghost_functor_outputs

### PR DESCRIPTION
**Pull request by lindsayad**
We appear to have a leak from this code as shown [here](https://civet.inl.gov/job/653571/). This patch appears to fix the issue as shown [here](https://civet.inl.gov/job/654071/).

### CODE REVIEW

**Comment by jwpeterson** [comment]
Nice catch. I haven't looked at this in detail, but it seems to me that `temporary_coupling_matrices` should store `unique_ptrs` so we can avoid calling `new`/`delete`. @roystgnr ?

**Comment by lindsayad** [comment]
I think @roystgnr is on vacation for a couple more weeks...?

**Comment by jwpeterson** [comment]
> I think @roystgnr is on vacation for a couple more weeks...?

OK, well this fix can definitely be merged as-is since it fixes the immediate problem, but it would be good to address the underlying issue of why we are manually managing memory here.

**Comment by fdkong** [comment]
Yes, we should use smart pointers because they are smart and valgrind-clean, lol

File: src/base/dof_map.C [hunk 1]
```diff
@@ -1483,7 +1483,13 @@ merge_ghost_functor_outputs(GhostingFunctor::map_type & elements_to_ghost,
...
                         {
                           CouplingMatrix * cm = new CouplingMatrix(*existing_it->second);
-                          temporary_coupling_matrices.insert(cm);
+#ifndef NDEBUG
+                          auto dbg_pr =
+#endif
+                            temporary_coupling_matrices.insert(cm);
+                          libmesh_assert_msg(dbg_pr.second,
+                                             "We must actually insert this new coupling matrix or "
+                                             "else we are going to leak");
```
Comment (roystgnr) [inline src/base/dof_map.C:1492]: How is this ever going to fail?  If we just got a pointer from new, then it won't alias any existing pointer and it won't be in the container.  If we didn't, then new should have thrown an exception and we shouldn't have made it to the assert.

File: src/base/dof_map.C [hunk 2]
```diff
@@ -1501,7 +1507,10 @@ merge_ghost_functor_outputs(GhostingFunctor::map_type & elements_to_ghost,
                       std::set<CouplingMatrix *>::iterator temp_it =
                         temporary_coupling_matrices.find(const_cast<CouplingMatrix *>(existing_it->second));
                       if (temp_it != temporary_coupling_matrices.end())
+                      {
+                        delete *temp_it;
```
Comment (roystgnr) [inline src/base/dof_map.C:1511]: Yeah, not sure how I missed this.

**Review by roystgnr — APPROVED** [review]

**Comment by roystgnr** [comment]
The code here was written before we started requiring C++11; otherwise it would have been smart pointers from the start.  We should definitely migrate to them.

**Comment by roystgnr** [comment]
And, yeah, on vacation until the 18th.

File: src/base/dof_map.C (hunk 1 above)
Comment (lindsayad) [reply src/base/dof_map.C:1492]: Yea...that seems pretty obvious when I give any level of thought about it

**Pull request merged**